### ISSUE TITLE
Realm file grows incredibly large on Realm 2.2.0+

### ISSUE THREAD

**Issue by raja-baz** [realm/realm-cocoa#4590]
Realm 2.2.0 seems to have introduced a bug somewhere that causes the realm file to grow exponentially under certain loads.

## Reproduction

```
#import "ViewController.h"
#import <Realm/RLMObject.h>
#import <Realm/RLMRealm.h>
#import <Realm/RLMRealmConfiguration.h>

@interface ViewController () {
    int runCount;
}

@end

@interface TestObject : RLMObject

@property int number;
@property NSString * data;

+ (TestObject *)randomTestObject;

@end

@interface TestThread : NSThread

@end

@implementation TestObject

+ (TestObject *)randomTestObject
{
    TestObject * result = [[TestObject alloc] init];
    result.number = arc4random_uniform(100);
    result.data = @"Hello world";
    return result;
}

+ (NSString *)primaryKey
{
    return @"number";
}

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    runCount = 0;
    RLMRealmConfiguration * config = [RLMRealmConfiguration defaultConfiguration];
    [[NSFileManager defaultManager] removeItemAtPath:config.fileURL.path error:nil];
    [self runTest];
}

+ (void)insertionRun
{
    @autoreleasepool {
        RLMRealm * realm = [RLMRealm defaultRealm];
        [realm transactionWithBlock:^{
            for(int i = 0; i < 1000; i++) {
                [realm addOrUpdateObject:[TestObject randomTestObject]];
            }
        }];
    }

}

- (NSString *)realmFileSize
{
    NSURL * fileUrl = [RLMRealm defaultRealm].configuration.fileURL;
    double size = [[NSFileManager defaultManager] attributesOfItemAtPath:fileUrl.path error:nil].fileSize;
    NSString * unit = @"";
    if (size > 1024) {
        unit = @"KB";
        size /= 1024.;
    }
    if (size > 1024) {
        unit = @"MB";
        size /= 1024;
    }
    return [NSString stringWithFormat:@"%.1f%@", size, unit];
}

- (void)runTest
{
    if (runCount % 100 == 0) {
        NSLog(@"%d %@", runCount, [self realmFileSize]);
    }
    if (runCount++ >= 1000) {
        NSLog(@"File size at end: %@", [self realmFileSize]);
        return;
    }
    for (int i = 0; i < 10; i++) {
        [[[TestThread alloc] init] start];
    }
    [self.class insertionRun];
    dispatch_async(dispatch_get_main_queue(), ^{
        [self runTest];
    });
    
}


@end

@implementation TestThread

- (void)main
{
    [ViewController insertionRun];
}

@end

```

I'm just hitting the db from main thread as well as background threads very hard. I make sure that Realm objects on the background threads get autoreleased properly.

### Realm 2.1.2 test run

```
2017-01-26 13:43:06.188 RealmSizeTest[42489:13113867] Version 2.3.0 of Realm is now available: https://github.com/realm/realm-cocoa/blob/v2.3.0/CHANGELOG.md
2017-01-26 13:43:06.199 RealmSizeTest[42489:13113760] 0 4.0KB
2017-01-26 13:43:16.264 RealmSizeTest[42489:13113760] 100 960.0KB
2017-01-26 13:43:26.912 RealmSizeTest[42489:13113760] 200 1.2MB
2017-01-26 13:43:37.363 RealmSizeTest[42489:13113760] 300 1.2MB
2017-01-26 13:43:48.082 RealmSizeTest[42489:13113760] 400 1.2MB
2017-01-26 13:43:58.873 RealmSizeTest[42489:13113760] 500 1.2MB
2017-01-26 13:44:09.543 RealmSizeTest[42489:13113760] 600 1.2MB
2017-01-26 13:44:20.290 RealmSizeTest[42489:13113760] 700 1.2MB
2017-01-26 13:44:31.305 RealmSizeTest[42489:13113760] 800 1.2MB
2017-01-26 13:44:42.534 RealmSizeTest[42489:13113760] 900 1.2MB
2017-01-26 13:44:52.024 RealmSizeTest[42489:13113760] 1000 1.2MB
2017-01-26 13:44:52.025 RealmSizeTest[42489:13113760] File size at end: 1.2MB
```

### Realm 2.2.0 test run

```
2017-01-26 13:47:59.049 RealmSizeTest[43088:13132384] Version 2.3.0 of Realm is now available: https://github.com/realm/realm-cocoa/blob/v2.3.0/CHANGELOG.md
2017-01-26 13:47:59.072 RealmSizeTest[43088:13132277] 0 4.0KB
2017-01-26 13:48:09.774 RealmSizeTest[43088:13132277] 100 2.8MB
2017-01-26 13:48:23.945 RealmSizeTest[43088:13132277] 200 48.0MB
2017-01-26 13:48:39.185 RealmSizeTest[43088:13132277] 300 120.0MB
2017-01-26 13:48:54.536 RealmSizeTest[43088:13132277] 400 208.0MB
2017-01-26 13:49:11.571 RealmSizeTest[43088:13132277] 500 352.0MB
2017-01-26 13:49:29.719 RealmSizeTest[43088:13132277] 600 480.0MB
2017-01-26 13:49:48.606 RealmSizeTest[43088:13132277] 700 640.0MB
2017-01-26 13:50:09.208 RealmSizeTest[43088:13132277] 800 832.0MB
2017-01-26 13:50:30.976 RealmSizeTest[43088:13132277] 900 1152.0MB
2017-01-26 13:50:54.418 RealmSizeTest[43088:13132277] 1000 1280.0MB
2017-01-26 13:50:54.418 RealmSizeTest[43088:13132277] File size at end: 1280.0MB
```

## Conclusion

Clearly, something happened in 2.2.0 that causes this. A test on 2.3.0 produces identical results. Note that not *every* test run triggers this behaviour(some do reach a steady state around 4MB) but most runs do. We ran into this bug on our app in regular when we updated from Realm 1.1.0 to 2.2.0. One of our test devices ended up with a Realm file it couldn't open and on investigation we found it was over 700MB big and had triggered the mmap() limitation.

As a side note, the steady state reached on the 1.x line is usually around 200~300KB. In the 2.x line it's 1~4MB. So there was an increase there, but at least it still reaches a steady state. On 2.2.0 it just keeps growing endlessly

**Comment by istx25** [comment]
Hey @raja-baz! Thanks for reporting this. We'll have someone review this and follow-up with you.

**Comment by samacs** [comment]
I think it would be worth it to advice developers in your documentation that due to the complexity of how Realm manages versions of data you will end up doing lots of `invalidate`'s and `refresh`'es all over the place. Not to mention wrapping every single access to a realm in an `@autorelease` block.

**Comment by samacs** [comment]
Is there any example out there showing how this can be avoided? @raja-baz @istx25 @jpsim

**Comment by bdash** [comment]
My initial impression from looking at the code you've provided is that you're being hurt by Realm's database write lock not being fair (that is, threads that block while attempting to take the lock are not unblocked in the order in which they attempted to take the lock). This can result in a thread waiting longer than expected to acquire the write lock, pinning the version of data it's looking at for the duration that it is blocked. We're tracking this underlying issue in realm/realm-core#2395. Until then you can mitigate this problem by performing your write transactions in serial rather than attempting many in parallel. There's no real reason to attempt to perform many writes in parallel as Realm only allows a single write transaction to be in progress at a time.

I'll attempt to reproduce using the code you've provided to confirm my hypothesis.

**Comment by raja-baz** [comment]
Maybe that's one of the contributing factors in this specific reproduction(which is obviously contrived so that it reproduces as fast as possible). But we hit this in production code that doesn't strain the db to anywhere near this degree.

We use Realm in our app(a music streaming app) for the following purposes:

1 - Image cache metadata(actual images stored on disk, but metadata about what is cached and sizes and stuff are in Realm).
2 - Song cache metadata(likewise, metadata about what's cached and stuff)
3 - Some data about recent plays
4 - Temporary storage of stuff that needs to be saved to the db pending receipt confirmation from our API(so that it can be resent if there's a problem).

That's about it, our usage patterns are not that heavy. In our production code, on Realm 2.2.0, it's enough to let the app stream for ~8hrs to hit Realm file sizes in the multiples of 100MB(where the size of the output of writeCopyToURL is about 200k). I highly doubt this is due to lock contention. So while locking may be a contributing factor, I'm sure it's not the main issue we're facing

**Comment by bdash** [comment]
It's going to be hard for us to determine what's going wrong if the sample code you provide doesn't reflect what your app is actually doing.

Would it be possible for you to share a prebuilt version of your app that reproduces the problem? That would allow us to observe the actual problem that you're seeing, and should make it possible for us to determine why you're seeing this file size growth. You can share the app with us via email to help@realm.io.

**Comment by raja-baz** [comment]
@bdash What, exactly would I share with you? An IPA of the app? Would you need me to sign it/provide provisioning profiles or can you manage that part?

As for the usage pattern in question. The reproduction code above hits at least *some* bug in 2.2.0 and I highly suspect that fixing one will fix the other(both our app and the code above have no issue on 2.1.2 and a reproducible issue on 2.2.0. So obviously something changed in that version)

**Comment by raja-baz** [comment]
One last thing, if my model of how the Realm file versions things and grows is correct, and what you're saying about the issue being locking is true. Wouldn't it follow that adding a `[realm invalidate]` at the end of the autorelease block in `insertRun` should have no effect? If we're being pinned while trying to perform a transaction and causing old copies to linger, then what happens afterwards should have no effect, correct?

However, if you invalidate the realm there, the issue goes away. My understanding of the matter is that the autorelease pool should have had the same effect, except it doesn't, for some reason

**Comment by samacs** [comment]
That's exactly what I ended up doing!

**Comment by bdash** [comment]
> @bdash What, exactly would I share with you? An IPA of the app? Would you need me to sign it/provide provisioning profiles or can you manage that part?

An IPA of the app is sufficient. I can then re-sign it to work on one of my devices. It would be ideal if the app is built using the dynamic Realm framework (rather than the static one) so that I can replace the contained framework to test fixes or add extra logging.

> As for the usage pattern in question. The reproduction code above hits at least some bug in 2.2.0 and I highly suspect that fixing one will fix the other(both our app and the code above have no issue on 2.1.2 and a reproducible issue on 2.2.0. So obviously something changed in that version)

What leads you to believe that the cause of the growth in the test code is the same as what you're experiencing in your real application? The test code is clearly prone to the lock contention / fairness issue that I described, while you say it's not likely that your app is hitting that.

> One last thing, if my model of how the Realm file versions things and grows is correct, and what you're saying about the issue being locking is true. Wouldn't it follow that adding a [realm invalidate] at the end of the autorelease block in insertRun should have no effect? If we're being pinned while trying to perform a transaction and causing old copies to linger, then what happens afterwards should have no effect, correct?

>However, if you invalidate the realm there, the issue goes away. My understanding of the matter is that the autorelease pool should have had the same effect, except it doesn't, for some reason

One thing I _did_ just notice is that your `-realmFileSize` method opens the default Realm in a way that will result in it ending up in the main thread's autorelease pool. It's possible that this instance is what results in an older version of the Realm data being pinned, and that the explicit `-invalidate` you're adding works to close that Realm instance's read transaction and unpin the older version within the Realm file.

**Comment by bdash** [comment]
I added logging of the initial and final version number within `realm::SharedGroup::promote_to_write`, which is what does the meat of the work of `-[RLMRealm beginWriteTransaction]`. I happened to reproduce the file size growth on my first run with the logging in place. What I saw is that in one single call to `promote_to_write`, the open Realm was ~1,500 versions behind the current (10-20 is typical for this test case). In instances where I see no file growth, the most behind a given Realm is at the time `promote_to_write` is called is < 30.

It's not yet clear _why_ the one call I saw lagged behind like that.

**Comment by bdash** [comment]
I wasn't able to reproduce that particular occurrence. Instead what I've seen since is that when the Realm file grows, it's because there's an `RLMRealm` instance created on a background thread whose lifetime is much longer than anticipated. I've been having trouble tracking down exactly why that is, but will resume looking into it tomorrow.

**Comment by bdash** [comment]
After looking into this further, I believe the problem is that `+[RLMRealm realmWithConfiguration:error:]`'s use of `RLMGetAnyCachedRealmForPath`results in a `RLMRealm` instance ending up in the autorelease pool of a different thread. This can result in that instance living much longer than expected, as repeated calls to `+[RLMRealm realmWithConfiguration:error]` may result in the lifetime of a single instance being repeatedly extended (if we're unlucky, and depending on how `NSMapTable` decides which object to return). If I enclose the call to `RLMGetAnyCachedRealmForPath` within an autorelease pool, limiting how long the `RLMRealm`'s lifetime is extended, I see the Realm file's final size consistently finish at < 1MB.

**Comment by bdash** [comment]
Some extra logging confirms that's what is happening.

**Comment by bdash** [comment]
I submitted a fix for that problem in #4598. In my testing it addresses the file size growth that occurs with the test code in this issue. It's difficult to say whether it'll help what you're seeing in your real app.
